packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -586,75 +590,73 @@ class TimelineFrame {
+  // Stores frame start time, end time, and duration.
+  final timeRange = TimeRange();
 
-  bool get isWellFormed =>
-      _pipelineItemStartTime != null && _pipelineItemEndTime != null;
+  /// Pipeline item time range in micros.
+  ///
+  /// This stores the start and end times for the pipeline item event for this
+  /// frame. We use this value to determine whether a TimelineEvent fits within
+  /// the frame's time boundaries.
+  final _pipelineItemTimeRange = TimeRange();
 
-  /// Frame start time in micros.
-  int get startTime => _uiStartTime;
+  int get pipelineItemStartTime => _pipelineItemTimeRange.start?.inMicroseconds;
 
-  /// Frame end time in micros.
-  int get endTime => _gpuEndTime;
+  set pipelineItemStartTime(int time) => _pipelineItemTimeRange.start =
+      Duration(microseconds: nullSafeMin(pipelineItemStartTime, time));
 
-  /// Duration the frame took to render in micros.
-  int get duration =>
-      endTime != null && startTime != null ? endTime - startTime : null;
+  int get pipelineItemEndTime => _pipelineItemTimeRange.end?.inMicroseconds;
 
-  // Timing info for UI portion of the frame.
-  int get _uiStartTime => uiEventFlow?.startTime;
+  set pipelineItemEndTime(int time) => _pipelineItemTimeRange.end =
+      Duration(microseconds: nullSafeMax(pipelineItemEndTime, time));
 
-  int get _uiEndTime => _uiStartTime != null ? _uiStartTime + uiDuration : null;
+  bool get isWellFormed =>
+      pipelineItemStartTime != null && pipelineItemEndTime != null;
 
-  int get uiDuration => uiEventFlow?.duration;
+  int get uiDuration => uiEventFlow != null
+      ? uiEventFlow.timeRange.duration.inMicroseconds
+      : null;
 
   double get uiDurationMs => uiDuration != null ? uiDuration / 1000 : null;
 
-  // Timing info for GPU portion of the frame.
-  int get _gpuStartTime => gpuEventFlow?.startTime;
+  int get gpuDuration => gpuEventFlow != null
+      ? gpuEventFlow.timeRange.duration.inMicroseconds
+      : null;
 
-  int get _gpuEndTime =>
-      _gpuStartTime != null ? _gpuStartTime + gpuDuration : null;
+  double get gpuDurationMs => gpuDuration != null ? gpuDuration / 1000 : null;
 
-  int get gpuDuration => gpuEventFlow?.duration;
+  void setEventFlow(TimelineEvent event, {TimelineEventType type}) {
+    type ??= event?.type;
 
-  double get gpuDurationMs => gpuDuration != null ? gpuDuration / 1000 : null;
+    eventFlows[type.index] = event;
+
+    if (type == TimelineEventType.ui) {
+      timeRange.start = event?.timeRange?.start;
+    }
+    if (type == TimelineEventType.gpu) {
+      timeRange.end = event?.timeRange?.end;
+    }
+  }
 
   @override
   String toString() {
-    return 'Frame $id - [start: $startTime], [end: $endTime],'
-        'ui: [start $_uiStartTime end $_uiEndTime], gpu: [start: $_gpuStartTime'
-        ' end $_gpuEndTime]';
+    return 'Frame $id - ${timeRange.toString()},'
+        ' ui: ${uiEventFlow.timeRange.toString()}, '
```

Comment:
you do not need the .toString().  String interpolation does that for you.

Reply:
removed.